src/common/logger/logger.ts

```diff
@@ -14,41 +14,112 @@
  * limitations under the License.
  */
 
+import { LoggerFormat } from "../../server/models/server-settings/server-settings";
 import { noop, Unary } from "../utils/functional/functional";
+import { isoNow } from "../utils/time/time";
+
+type LogFn = (msg: string, extra?: Record<string, string>) => void;
 
 export interface Logger {
-  log: Function;
-  warn: Function;
-  error: Function;
-  addPrefix: Unary<String, Logger>;
+  log: LogFn;
+  warn: LogFn;
+  error: LogFn;
+  setLoggerId: Unary<String, Logger>;
+}
+
+class JSONLogger implements Logger {
+
+  constructor(private logger = "turnilo") {
+  }
+
+  log(message: string, extra: Record<string, string> = {}) {
+    console.log(JSON.stringify({
+      message,
+      time: isoNow(),
+      level: "INFO",
+      logger: this.logger,
+      ...extra
+    }));
+  }
+
+  setLoggerId(loggerId: string): Logger {
+    return new JSONLogger(loggerId);
+  }
+
+  error(message: string, extra: Record<string, string> = {}) {
+    console.log(JSON.stringify({
+      message,
+      time: isoNow(),
+      level: "ERROR",
+      logger: this.logger,
+      ...extra
+    }));
+  }
+
+  warn(message: string, extra: Record<string, string> = {}) {
+    console.log(JSON.stringify({
+      message,
+      time: isoNow(),
+      level: "WARN",
+      logger: this.logger,
+      ...extra
+    }));
+  }
 }
 
 class ConsoleLogger implements Logger {
-  constructor(private prefixes: string[] = []) {
+  constructor(private prefix = "") {
   }
 
-  error(...args: any[]) {
-    console.error(...this.prefixes, ...args);
+  error(message: string) {
+    console.error(this.prefix, message);
   }
 
-  warn(...args: any[]) {
-    console.warn(...this.prefixes, ...args);
+  warn(message: string) {
+    console.warn(this.prefix, message);
   }
 
-  log(...args: any[]) {
-    console.log(...this.prefixes, ...args);
+  log(message: string) {
+    console.log(this.prefix, message);
   }
 
-  addPrefix(prefix: string): Logger {
-    return new ConsoleLogger([...this.prefixes, prefix]);
+  setLoggerId(loggerId: string): Logger {
+    return new ConsoleLogger(loggerId);
   }
 }
 
-export const LOGGER: Logger = new ConsoleLogger();
+class ErrorLogger implements Logger {
```

Comment:
Logger used for introspection. We want to redirect all logs into stderr, so it won't interfere with produced config that's directed to stdout.

Reply:
ConsoleAlwaysErrorLogger :)